### PR TITLE
Add basic Guest play support

### LINKED ISSUE #106: Enter as Guest is not working
Enter as guest is not working, maybe due to button is not mapped to any action.

### PR DESCRIPTION
**Pull request by N4r35h**
[basic_guest_chess.webm](https://github.com/code100x/chess/assets/45840495/d38306d9-cfae-47f7-865e-ff4f3d7e6657)

just a very basic implementation that gets guest play working 
- guest vs guest (no db saves)
- user vs guest (no db saves)
- user vs user

if game contains a guest we wont save to the database this is because the guest is not a valid db user we might make a default user for that and save so in case of user vs guest scenario we can show the user his history (if shown in the future)

this is meant to kick things off 

please share your inputs on this @hkirat and others

resolves #106

### CODE REVIEW

**Review by malharrrr — APPROVED** [review]

File: apps/ws/src/Game.ts [hunk 1]
```diff
@@ -30,46 +31,56 @@ export function isPromoting(chess: Chess, from: Square, to: Square) {
...
-        } catch(e) {
-            console.error(e)
-            return;
+    async updateSecondPlayer(player2: Player) {
+        this.player2 = player2;
+        let player1Name = this.player1.name
+        let player2Name = this.player2?.name
+
+        // only create the game in db if its not a guest game
+        if (!this.isGuestGame()) {
```
Comment (hkirat) [inline apps/ws/src/Game.ts:69]: Still needs to be in Db
1. create a new table called guests
2. Can we somehow have an `or` on foreign keys from Game  (whiteplayer, blackPlayer) to Users | Guests (answer could totally be no)

**Comment by N4r35h** [comment]
@hkirat if both players are guest no need to persist to db right ? it really serves no purpose

File: apps/ws/src/Game.ts (hunk 1 above)
Comment (nimit9) [reply apps/ws/src/Game.ts:69]: Maybe we can have a GuestGame table which points to GuestMoves table?

**Comment by nimit9** [comment]
> @hkirat if both players are a guest no need to persist to db right ? it really serves no purpose 
> can we have this merged and take up persisting in another pr ? this at-least gets guest play kick started
> 
> if any one user is guest can we just point that to a "Default precreated guest user" ? 
> also please explain more on what we will perform saving the guests
> 
> share your thoughts

I think it's mostly needed, because you want to recover the game, if the user get's disconnected by mistake/internet issue, or refreshes etc.

**Comment by hkirat** [comment]
we also need to allow them to share their games and replay it

**Comment by hkirat** [comment]
We can just add guests in the `User` table for now with `type` set to `guest`

**Comment by N4r35h** [comment]
> @N4r35h let me know if u'd like to pursue this further, if not I'll merge some parts of this and finish the rest

I would like to pursue this just a few confirmations

Even if both players are guest we want to store to db ?
We will have a user entry for each guest ?

**Comment by N4r35h** [comment]
@hkirat now we save guests to the database and persist the games

> We can just add guests in the `User` table for now with `type` set to `guest`

have a look and if it looks good merge it in

one known issue that i tried to but haven't solved yet is to persist guest sessions in the express session or in cookie

i wouldn't be available for the rest of the day pretty much so feel free to merge and fix if you wish

**Comment by N4r35h** [comment]
@hkirat now we save guests to the database and persist the games

> We can just add guests in the `User` table for now with `type` set to `guest`

have a look and if it looks good merge it in

guest sessions are done via a separate jwt in a cookie called `guest` so if a guest refreshes the page or for example looses internet connection for a few seconds he can still rejoin and play 

let me know if this seperate cookie route would be ok or if we have to hack it into the passport.js way not sure if that would be possible maybe with some weird custom logic like a GuestAuthProvider idk

File: apps/backend/src/router/auth.ts [hunk 2]
```diff
@@ -2,16 +2,55 @@ import { Request, Response, Router } from 'express';
...
+    { userId: user.id, name: user.name, isGuest: true },
+    JWT_SECRET,
+  );
+  let User: User = {
+    id: user.id,
+    name: user.name!,
+    token: token,
+    isGuest: true,
+  };
+  res.cookie('guest', token, { maxAge: COOKIE_MAX_AGE });
```
Comment (hkirat) [inline apps/backend/src/router/auth.ts:50]: return to user and store in local storage

Comment (hkirat) [reply apps/backend/src/router/auth.ts:50]: nvm ignore this

File: apps/frontend/src/screens/Game.tsx [hunk 3]
```diff
@@ -95,6 +101,9 @@ export const Game = () => {
           setResult(message.payload.result);
           break;
 
+        case OPPONENT_DISCONNECTED:
+          setResult(OPPONENT_DISCONNECTED);
```
Comment (hkirat) [inline apps/frontend/src/screens/Game.tsx:105]: Is this needed?

File: apps/ws/src/Game.ts [hunk 4]
```diff
@@ -71,12 +72,14 @@ export class Game {
         payload: {
           gameId: this.gameId,
           whitePlayer: {
-            name: users.find((user) => user.id === this.player1UserId)?.name,
```
Comment (hkirat) [inline apps/ws/src/Game.ts:74]: Why cant we ues the DB entry? 
Why are we propogating hte name to Game.ts?

File: apps/ws/src/GameManager.ts [hunk 5]
```diff
@@ -51,7 +51,7 @@ export class GameManager {
             console.error('Pending game not found?');
             return;
           }
-          if (user.userId === game.player1UserId) {
+          if (user.userId === game.player1.userId) {
```
Comment (hkirat) [inline apps/ws/src/GameManager.ts:54]: revert this as u revert the Game.ts logic

File: apps/ws/src/GameManager.ts [hunk 6]
```diff
@@ -118,8 +118,14 @@ export class GameManager {
 
         if (!availableGame) {
           const game = new Game(
-            gameFromDb?.whitePlayerId!,
-            gameFromDb?.blackPlayerId!,
+            {
+              name: gameFromDb?.whitePlayer?.name || '',
```
Comment (hkirat) [inline apps/ws/src/GameManager.ts:122]: dont think this is needed

**Review by hkirat — CHANGES_REQUESTED** [review]
some minor comments 
mostly around whats happening in `Game.ts`
Could u check them out please?

File: apps/backend/src/router/auth.ts (hunk 2 above)
Comment (N4r35h) [reply apps/backend/src/router/auth.ts:50]: ok ignoring this, sticking to guest cookie, resolving this

File: apps/frontend/src/screens/Game.tsx (hunk 3 above)
Comment (N4r35h) [reply apps/frontend/src/screens/Game.tsx]: oops got it duplicated while merging, removed duplicate

File: apps/ws/src/Game.ts (hunk 4 above)
Comment (N4r35h) [reply apps/ws/src/Game.ts:86]: yes was previous logic to handle guest names when guests where not stored in database, reverted

File: apps/ws/src/GameManager.ts (hunk 5 above)
Comment (N4r35h) [reply apps/ws/src/GameManager.ts]: reverted

File: apps/ws/src/GameManager.ts (hunk 6 above)
Comment (N4r35h) [reply apps/ws/src/GameManager.ts]: reverted as well

**Comment by N4r35h** [comment]
@hkirat made the requested changes please have a look now

**Review by malharrrr — COMMENTED** [review]
looks good

**Comment by N4r35h** [comment]
@hkirat should be good to merge, resolved conflicts

**Comment by N4r35h** [comment]
@hkirat could you please take a look at this PR ?

File: apps/backend/src/router/auth.ts [hunk 7]
```diff
@@ -2,16 +2,55 @@ import { Request, Response, Router } from 'express';
...
+      name: bodyData.name || guestUUID,
+      provider: 'GUEST',
+    },
+  });
+
+  const token = jwt.sign(
+    { userId: user.id, name: user.name, isGuest: true },
+    JWT_SECRET,
+  );
+  let User: User = {
```
Comment (siinghd) [inline apps/backend/src/router/auth.ts:44]: Whats going on here? can we change this?
use const and change User to something else

File: apps/backend/src/index.ts [hunk 8]
```diff
@@ -6,16 +6,21 @@ import authRoute from './router/auth';
 import dotenv from 'dotenv';
 import session from 'express-session';
 import passport from 'passport';
+import cookieParser from 'cookie-parser';
+
+export const COOKIE_MAX_AGE = 24 * 60 * 60 * 1000;
```
Comment (siinghd) [inline apps/backend/src/index.ts:11]: maybe create a seperate file to store constats?

File: apps/ws/src/Game.ts [hunk 9]
```diff
@@ -124,19 +125,24 @@ export class Game {
       return;
     }
 
+    let WhitePlayer = users.find((user) => user.id === this.player1UserId);
```
Comment (siinghd) [inline apps/ws/src/Game.ts:128]: also here maybe const whitePlayer etc...

File: apps/ws/src/index.ts [hunk 10]
```diff
@@ -11,8 +10,8 @@ const gameManager = new GameManager();
 wss.on('connection', function connection(ws, req) {
   //@ts-ignore
   const token: string = url.parse(req.url, true).query.token;
-  const userId = extractUserId(token);
-  gameManager.addUser(new User(ws, userId));
+  const User = extractAuthUser(token, ws);
```
Comment (siinghd) [inline apps/ws/src/index.ts:13]: lets stick with lower case first letter for vars

**Review by siinghd — CHANGES_REQUESTED** [review]
Resolve the merge conflicts and these tiny things and we are good to go

File: apps/backend/src/router/auth.ts (hunk 7 above)
Comment (N4r35h) [reply apps/backend/src/router/auth.ts:44]: we are returning userDetails so we can store it in the frontend state, changed the struct and variable names from "User" to "UserDetails"

File: apps/backend/src/index.ts (hunk 8 above)
Comment (N4r35h) [reply apps/backend/src/index.ts:11]: moved to a separate file

File: apps/ws/src/Game.ts (hunk 9 above)
Comment (N4r35h) [reply apps/ws/src/Game.ts:128]: changed those variables to constants

File: apps/ws/src/index.ts (hunk 10 above)
Comment (N4r35h) [reply apps/ws/src/index.ts:13]: changed to camel case

**Review by malharrrr — APPROVED** [review]

File: apps/backend/package.json [hunk 11]
```diff
@@ -13,7 +13,9 @@
   "license": "ISC",
   "dependencies": {
     "@repo/db": "*",
+    "@types/cookie-parser": "^1.4.7",
```
Comment (siinghd) [inline apps/backend/package.json:16]: these should be dev dependencies right?

**Review by siinghd — CHANGES_REQUESTED** [review]

File: apps/backend/package.json (hunk 11 above)
Comment (N4r35h) [reply apps/backend/package.json]: yes, updated and moved it from dep to devDep

**Comment by N4r35h** [comment]
@hkirat any bounty for this ?

**Pull request merged**